cds_ils/importer/documents/importer.py

```diff
@@ -0,0 +1,110 @@
+import uuid
+from copy import deepcopy
+
+import click
+from invenio_app_ils.documents.api import DocumentIdProvider
+from invenio_app_ils.errors import IlsValidationError
+from invenio_app_ils.proxies import current_app_ils
+from invenio_db import db
+
+from cds_ils.importer.documents.api import search_documents_by_isbn, \
+    search_documents_by_doi, get_document_by_exact_title_authors
+
+
+class DocumentImporter(object):
+
+    def __init__(self, json_metadata, helper_metadata_fields,
+                 metadata_provider, update_document_fields):
+        self.helper_metadata_fields = helper_metadata_fields
+        self.json_data = json_metadata
+        self.metadata_provider = metadata_provider
+        self.update_document_fields = update_document_fields
+
+    def _set_record_import_source(self, record_dict):
+        record_dict["created_by"] = {"type": "import",
+                                     "value": self.metadata_provider}
+
+    def _before_create(self):
+        cleaned = deepcopy(self.json_data)
+        for field in self.helper_metadata_fields:
+            if field in cleaned:
+                del cleaned[field]
+        # save the import source
+        self._set_record_import_source(cleaned)
+        return cleaned
+
+    def create_document(self):
+        """Create a new record from dump."""
+        cleaned_json = self._before_create()
+        document_class = current_app_ils.document_record_cls
+        # Reserve record identifier, create record and recid pid in one
+        # operation.
+        record_uuid = uuid.uuid4()
+        provider = DocumentIdProvider.create(
+            object_type="rec",
+            object_uuid=record_uuid,
+        )
+        cleaned_json["pid"] = provider.pid.pid_value
+
+        try:
+            document = document_class.create(cleaned_json, record_uuid)
+            db.session.commit()
+            return document
+        except IlsValidationError as e:
+            click.secho("Field: {}".format(e.errors[0].res["field"]), fg="red")
```

Comment:
do we have to `rollback` here too? Otherwise the PID is almost minted, but not the doc... or maybe:
```
with db.session.nested:
  PID
  Record
db.session.commit
or 
rolllback
```

Reply:
maybe you have missed this comment?